src/Build/Graph/ProjectGraph.cs

```diff
@@ -219,47 +218,100 @@ public ProjectGraph(string entryProjectFile)
-            while (projectsToEvaluate.Count != 0)
+            nodeState[projectToEvaluate] = NodeState.InProcess;
+            ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
+            IEnumerable<ProjectItemInstance> projectReferenceItems =
+                parsedProject.Project.GetItems(ProjectReferenceItemName);
+            foreach (var projectReferenceToParse in projectReferenceItems)
             {
-                KeyValuePair<ConfigurationMetadata, ProjectGraphNode> projectToEvaluateAndParentNode = projectsToEvaluate.Dequeue();
-                ConfigurationMetadata projectToEvaluate = projectToEvaluateAndParentNode.Key;
-                ProjectGraphNode parentNode = projectToEvaluateAndParentNode.Value;
+                if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                {
+                    throw new InvalidOperationException(string.Format(
+                        CultureInfo.InvariantCulture,
+                        ResourceUtilities.GetResourceString(
+                            "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                        projectReferenceToParse.EvaluatedInclude,
+                        parsedProject.Project.FullPath));
+                }
 
-                if (!_allParsedProjects.TryGetValue(projectToEvaluate, out ProjectGraphNode parsedProject))
+                string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
+                PropertyDictionary<ProjectPropertyInstance> projectReferenceGlobalProperties =
+                    GetProjectReferenceGlobalProperties(projectReferenceToParse, globalProperties);
+                var projectReferenceConfigurationMetadata =
+                    new ConfigurationMetadata(projectReferenceFullPath, projectReferenceGlobalProperties);
+                if (nodeState.TryGetValue(projectReferenceConfigurationMetadata, out NodeState projectReferenceNodeState))
                 {
-                    parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
-                    IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.Project.GetItems(ProjectReferenceItemName);
-                    foreach (var projectReferenceToParse in projectReferenceItems)
+                    // a project reference can be in "Processed" state. If it is "InProcess" state, it is an ancestor and there is a circular dependency
+                    if (projectReferenceNodeState == NodeState.InProcess)
                     {
-                        if (!string.IsNullOrEmpty(projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName)))
+                        if (projectToEvaluate.ProjectFullPath.Equals(projectReferenceConfigurationMetadata.ProjectFullPath))
```

Comment:
`ConfigurationMetadata` is equatable, so I think you should just call `projectToEvaluate.Equals(projectReferenceConfigurationMetadata)`